File: 2212_twoway/processing/godual_ranging.m
pkg load signal
% graphics_toolkit('gnuplot')

fs=5e6;
Nint=1;

dirlist=dir('./*bin')
dirbit=dir('./code/*bin')
for dirnum=1:length(dirlist)
  dirbit(mod(dirnum-1,length(dirbit))+1).name
  f=fopen(['code/',dirbit(mod(dirnum-1,length(dirbit))+1).name]);
  code=fread(f,inf,'int8');
  code=repelems(code,[[1:length(code)] ; ones(1,length(code))*2]); % interpolate
  code=2*code-1;  % +1/-1
  fcode=conj(fft(code'));
  fclose(f);

  dirlist.name
  f=fopen(dirlist(dirnum).name);
  
  p=1;
  do
    d=fread(f,length(fcode)*4,'int16');  % 1 code length
    longueur=length(d);
    d=d(1:2:end)+j*d(2:2:end);
    d1=d(1:2:end);  % measurement
    d2=d(2:2:end);  % reference
    d1=d1-mean(d1);
    d2=d2-mean(d2);
    clear d
  %  x1avant=abs(xcorr(d1,code));x1avant=x1avant(length(code):end);
    if (longueur==length(fcode)*4)  
      freq=linspace(-fs/2,fs/2,length(d1));
      k=find((freq<106200)&(freq>-96200));
  %%% d1
      d22=fftshift(abs(fft(d1.^2))); % 0.1 Hz accuracy
      [~,df1(p)]=max(d22(k));df1(p)=df1(p)+k(1)-1;df1(p)=freq(df1(p))/2;offset1=df1(p);
      temps=[0:length(d1)-1]'/fs;
      lo=exp(-j*2*pi*df1(p)*temps); % frequency offset
    
      y=d1.*lo;                      % frequency transposition
  % the phase of the correlation can be found to identify the oscillator drift but not before correlation at low SNR!
      prnmap01=fftshift(fft(y).*fcode);      % xcorr
      if (p==fs/length(code)*10) % after 10 s
         figure
         plot(abs(ifft(fftshift(prnmap01))))
         xlabel('offset (index)');ylabel('xcorr (a.u.)')
      end
      prnmap01=[zeros(length(y)*(Nint),1) ; prnmap01 ; zeros(length(y)*(Nint),1)]; % interpolation to 3x samp_rate
      prnmap01=(ifft(fftshift(prnmap01)));       % back to time /!\ NO outer fftshift for 0-delay at left
      [~,indice1(p)]=max(abs(prnmap01));
      xval1(p)=prnmap01(indice1(p));
      xval1m1(p)=prnmap01(indice1(p)-1);
      xval1p1(p)=prnmap01(indice1(p)+1);
      [u,v]=polyfit([-1:+1]',abs(prnmap01([indice1(p)-1:indice1(p)+1])),2);
      correction1_1(p)=-u(2)/2/u(1);
      [u,v]=polyfit([-2:+2]',abs(prnmap01([indice1(p)-2:indice1(p)+2])),2);
      correction1_2(p)=-u(2)/2/u(1);
      [u,v]=polyfit([-3:+3]',abs(prnmap01([indice1(p)-3:indice1(p)+3])),2);
      correction1_3(p)=-u(2)/2/u(1);
      correction1_a(p)=(abs(prnmap01(indice1(p)-1))-abs(prnmap01(indice1(p)+1)))/(abs(prnmap01(indice1(p)-1))+abs(prnmap01(indice1(p)+1))-2*abs(prnmap01(indice1(p))))/2;
  % SNR computation
      yf=fftshift(fft(y));
      yint=[zeros(length(y)*(Nint),1) ; yf ; zeros(length(y)*(Nint),1)]; % interpolation to 3x samp_rate
      yint=(ifft(fftshift(yint)));       % back to time /!\ outer fftshift for 0-delay at center
      codetmp=repelems(code,[[1:length(code)] ; ones(1,length(code))*(2*Nint+1)])'; % interpolate
      yincode=[yint(indice1(p)-1:end) ; yint(1:indice1(p)-2)].*codetmp;
      SNR1r(p)=mean(real(yincode))^2/var(yincode);
      SNR1i(p)=mean(imag(yincode))^2/var(yincode);
  %    cf=fftshift(fft(code));
  %    cint=[zeros(length(code)*(Nint),1) ; cf.' ; zeros(length(code)*(Nint),1)]; % interpolation to 3x samp_rate
  %    cint=real((ifft(fftshift(cint))));       % back to time /!\ outer fftshift for 0-delay at center
  
  %%% d2
      d22=fftshift(abs(fft(d2.^2))); % 0.1 Hz accuracy
      [~,df2(p)]=max(d22(k));df2(p)=df2(p)+k(1)-1;df2(p)=freq(df2(p))/2;offset2=df2(p);
      temps=[0:length(d2)-1]'/fs;
      lo=exp(-j*2*pi*df2(p)*temps); % frequency offset
    
      y=d2.*lo;                      % frequency transposition
      prnmap02=fftshift(fft(y).*fcode);      % xcorr
      prnmap02=[zeros(length(y)*(Nint),1) ; prnmap02 ; zeros(length(y)*(Nint),1)]; % interpolation to 3x samp_rate
      prnmap02=(ifft(fftshift(prnmap02)));       % back to time
      [~,indice2(p)]=max(abs(prnmap02));
      xval2(p)=prnmap02(indice2(p));
      xval2m1(p)=prnmap02(indice2(p)-1);
      xval2p1(p)=prnmap02(indice2(p)+1);
      [u,v]=polyfit([-1:+1]',abs(prnmap02([indice2(p)-1:indice2(p)+1])),2);
      correction2_1(p)=-u(2)/2/u(1);
      [u,v]=polyfit([-2:+2]',abs(prnmap02([indice2(p)-2:indice2(p)+2])),2);
      correction2_2(p)=-u(2)/2/u(1);
      [u,v]=polyfit([-3:+3]',abs(prnmap02([indice2(p)-3:indice2(p)+3])),2);
      correction2_3(p)=-u(2)/2/u(1);
      correction2_a(p)=(abs(prnmap02(indice2(p)-1))-abs(prnmap02(indice2(p)+1)))/(abs(prnmap02(indice2(p)-1))+abs(prnmap02(indice2(p)+1))-2*abs(prnmap02(indice2(p))))/2;
  % SNR computation
      yf=fftshift(fft(y));
      yint=[zeros(length(y)*(Nint),1) ; yf ; zeros(length(y)*(Nint),1)]; % interpolation to 3x samp_rate
      yint=(ifft(fftshift(yint)));       % back to time /!\ outer fftshift for 0-delay at center
      yincode=[yint(indice2(p)-1:end) ; yint(1:indice2(p)-2)].*codetmp;
      SNR2r(p)=mean(real(yincode))^2/var(yincode);
      SNR2i(p)=mean(imag(yincode))^2/var(yincode);
      if (p==fs/length(code)*10) % after 10 s
         figure
         plot(angle([yint(indice2(p)-1:end) ; yint(1:indice2(p)-2)]))
         xlim([0 1200])
         hold on
         plot(codetmp)
         pause(0.1)
      end
      printf("%d\t%f\t%f\t%f\t%f\t%f\r\n",p,offset1,10*log10(SNR1i(p)+SNR1r(p)),offset2,10*log10(SNR2i(p)+SNR2r(p)),(indice1(p)+correction1_a(p)-indice2(p)-correction2_a(p))/fs/(2*Nint+1))
      p=p+1;
    end
  until (longueur<length(fcode)*4);
  fclose(f)
  solution1=indice1+correction1_a;
  solution2=indice2+correction2_a;
  subplot(211)
  plot((solution1)/(2*Nint+1)/fs); % ranging solution
  xlabel('time (s)')
  ylabel('ranging delay (s)')
  [a,b]=polyfit([1:length(solution1)],(solution1)/(2*Nint+1)/fs,2);
  subplot(212)
  plot((solution1)/(2*Nint+1)/fs-b.yf); % ranging solution
  std((solution1)/(2*Nint+1)/fs-b.yf)
  mean((solution1)/(2*Nint+1)/fs-b.yf)
  legend(num2str(std((solution1)/(2*Nint+1)/fs-b.yf)))
  xlabel('time (s)')
  ylabel('delay - parabolic fit (s)')
  nom=strrep(dirlist(dirnum).name,'.bin','.mat');
  eval(['save -mat /tmp/',nom,' xv* corr* df1 df2 indic* SNR* code']);
  clear xv* corr* df* indic* p SNR*
end

File: 221207_twoway_codes/processing/godual_ranging.py
#!/usr/bin/env python3

import numpy as np
import struct
import array
from datetime import datetime

fs = (5e6)
foffset=0
frange=8000
Nint=1
affiche=0
debug=0

if (affiche==1):
    import matplotlib.pyplot as plt

def ranging(filename, prn_code):
    print(filename[-14:-4])
    ts = int(filename[-14:-4])

    with open(prn_code, "rb") as fd:
        codeb = fd.read()
        print(codeb[0:20])
        code = list(codeb)
        code=np.repeat(code,2)  # interpolate
        print(code[0:40])
        code=code*2-1
        fcode=np.conj(np.fft.fft(code))
    freq = np.linspace(-fs/2, (fs/2), num=len(code), dtype=float)
    k=np.nonzero((np.array(freq) < 2*(foffset+frange)) & (np.array(freq) > 2*(foffset-frange)))
    k=k[0];
    temps = np.array(range(0, len(code))) / fs
    p = 0
    df = []
    indice1_lst = []
    indice2_lst = []
    SNR1r_lst = []
    SNR1i_lst = []
    SNR2r_lst = []
    SNR2i_lst = []
    interpolation1=np.zeros((Nint*2+1)*len(code))+1j*np.zeros((Nint*2+1)*len(code))  # prepare empty array for interpolation
    interpolation2=np.zeros((Nint*2+1)*len(code))+1j*np.zeros((Nint*2+1)*len(code))
    yint=np.zeros((2*Nint+1)*len(code))+1j*np.zeros((2*Nint+1)*len(code))
    with open(filename, "rb") as fd:
        must_stop = False
        while(not must_stop):
            data = fd.read(len(code) * 4 * 2)
            if len(data) != len(code) * 4 * 2:
                break
            d = struct.unpack('{}h'.format(len(data)//2),data)
            d1 = np.array(d[0::4], dtype=complex)
            d1.imag = np.array(d[1::4])
            d2 = np.array(d[2::4], dtype=complex)
            d2.imag = np.array(d[3::4])
            d1 -= np.mean(d1)
            d2 -= np.mean(d2)

            d1_fft = np.fft.fftshift(np.abs(np.fft.fft(d1 * d1)))
            tmp = d1_fft[k].argmax()+k[0]
            dftmp = freq[tmp]/2
            df.append(tmp)
            lo = np.exp(-1j * 2 * np.pi * dftmp * temps)
            y = d1 * lo              # frequency transposition
            fft1tmp=np.fft.fft(y)
            fft2tmp=np.fft.fft(d2)
            multmp1 = fft1tmp * fcode
            multmp2 = fft2tmp * fcode
#            interpolation1=np.concatenate( (np.zeros(len(y))+1j*np.zeros(len(y)) , multmp1, np.zeros(len(y))+1j*np.zeros(len(y))))     # Nint=1
            interpolation1[:len(y)//2]=multmp1[:len(y)//2]
            interpolation1[-len(y)//2:]=multmp1[-len(y)//2:]
#            interpolation2=np.concatenate( (np.zeros(len(d2))+1j*np.zeros(len(d2)) , multmp2, np.zeros(len(d2))+1j*np.zeros(len(d2)))) # Nint=1
            interpolation2[:len(y)//2]=multmp2[:len(y)//2]
            interpolation2[-len(y)//2:]=multmp2[-len(y)//2:]
#            multmp1 = np.fft.fftshift(interpolation1)
#            multmp2 = np.fft.fftshift(interpolation2)
            prnmap01 = np.fft.ifft(interpolation1)       # correlation with returned signal
            prnmap02 = np.fft.ifft(interpolation2)

            indice1 = abs(prnmap01).argmax()      # only one correlation peak
            indice2 = abs(prnmap02).argmax()
            xval1 = prnmap01[indice1]
            xval2 = prnmap02[indice2]
            xval1m1= prnmap01[indice1-1]
            xval1p1 = prnmap01[indice1+1]
            xval2m1 = prnmap02[indice2-1]
            xval2p1 = prnmap02[indice2+1]
            if ( (p==1) and (affiche==1)):
                plt.figure()
                plt.plot(abs(prnmap01[indice1-2:indice1+3]))
                plt.plot(abs(prnmap02[indice2-2:indice2+3]))
                plt.show()
            correction1=((abs(xval1m1)-abs(xval1p1))/(abs(xval1m1)+abs(xval1p1)-2*abs(xval1))/2);
            correction2=((abs(xval2m1)-abs(xval2p1))/(abs(xval2m1)+abs(xval2p1)-2*abs(xval2))/2);
            
  # SNR1 computation
#            yf = np.fft.fftshift(np.fft.fft(y))
#            yint = np.concatenate( (np.zeros(len(y))+1j*np.zeros(len(yf)) , yf , np.zeros(len(y))+1j*np.zeros(len(y))))  # Nint=1
#            yint=np.fft.ifft(np.fft.fftshift(yint))                     # back to time /!\ outer fftshift for 0-delay at center
            yint[:len(y)//2]=fft1tmp[:len(y)//2]
            yint[-len(y)//2:]=fft1tmp[-len(y)//2:]
            yinti=np.fft.ifft(yint) 
            codetmp=np.repeat(code,2*Nint+1)   # interpolate 2*Nint+1
            yincode=np.concatenate((yinti[indice1-2:] , yinti[:indice1-2]))*codetmp; # -2 JMF CHECK WHY -2
            SNR1r=np.mean(np.real(yincode))**2/np.var(yincode);
            SNR1i=np.mean(np.imag(yincode))**2/np.var(yincode);
            puissance1=np.var(y)

  # SNR2 computation
            #yf = np.fft.fftshift(np.fft.fft(d2))
            #yint = np.concatenate( (np.zeros(len(y))+1j*np.zeros(len(yf)) , yf , np.zeros(len(y))+1j*np.zeros(len(y))))  # Nint=1
            #yint=np.fft.ifft(np.fft.fftshift(yint))                     # back to time /!\ outer fftshift for 0-delay at center
            yint[:len(y)//2]=fft2tmp[:len(y)//2]
            yint[-len(y)//2:]=fft2tmp[-len(y)//2:]
            yinti=np.fft.ifft(yint) 
            # if ( (p==fs/len(code)*10) and (affiche==1)):
            if ( (p==1) and (affiche==1)):
                plt.figure()
                plt.plot(np.real(np.concatenate( (yinti[indice2-2:] , yinti[:indice2-2]) ))[0:1001]/max(np.real(yinti[indice2-2:])))
                plt.plot(codetmp[0:1001])
                plt.show()
            yincode=np.concatenate((yinti[indice2:] , yinti[:indice2]))*codetmp; # -2 JMF CHECK WHY -2
            SNR2r=np.mean(np.real(yincode))**2/np.var(yincode);
            SNR2i=np.mean(np.imag(yincode))**2/np.var(yincode);
            puissance2=np.mean(np.real(yincode))**2+np.mean(np.imag(yincode))**2
            indice1_lst.append(indice1+correction1)
            indice2_lst.append(indice2+correction2)
            SNR1r_lst.append(SNR1r)
            SNR1i_lst.append(SNR1i)
            SNR2r_lst.append(SNR2r)
            SNR2i_lst.append(SNR2i)
# year month day hour minute second delay frequency power SNR
            if (debug == 1):
                print(str(p)+": "+str(indice1)+" "+str(correction1)+" "+str(indice2)+" "+str(correction2)+" "+str(round(abs(xval1m1),0))+" "+str(round(abs(xval1),0))+" "+str(round(abs(xval1p1),0))+" "+str(round(abs(xval2m1),0))+" "+str(round(abs(xval2),0))+" "+str(round(abs(xval2p1),0)))
            print(str(p)+": "+datetime.utcfromtimestamp(ts+p).strftime('%Y %m %d %H %M %S')+"\t"+str(round((indice1-indice2+correction1-correction2)/fs/3,12))+"\t"+str(round(dftmp,1))+"\t"+str(round(10*np.log10(puissance1),1))+"\t"+str(round(10*np.log10(SNR1i+SNR1r),1))+"\t"+str(round(10*np.log10(SNR2i+SNR2r),1)))
            p += 1

ranging("../tmp/1670074501.bin","../codes/noiselen500000_bitlen19_taps39.bin");
